echopype/clean/utils.py

```diff
@@ -83,6 +84,95 @@ def calc_transient_noise_pooled_Sv(
     return pooled_Sv
 
 
+def index_binning_pool_Sv(
+    ds_Sv: xr.Dataset,
+    func: str,
+    depth_bin: int,
+    num_side_pings: int,
+    exclude_above: float,
+    chunk_dict: dict,
+) -> xr.DataArray:
+    """
+    Compute pooled Sv array for transient noise masking using index binning.
+
+    This function makes the assumption that within each channel, the difference
+    between depth values is uniform across all pings. Thus, computing the number of
+    range sample indices needed to cover the depth bin is a channel-specific task.
+    """
+    # Create `ds_Sv` copy
+    ds_Sv_copy = (
+        ds_Sv.copy()
+        # Drop `filenames` dimension if exists and transpose Dataset
+        .drop_dims("filenames", errors="ignore").transpose("channel", "ping_time", "range_sample")
+    )
+
+    # Compute number of range sample indices that are needed to encapsulate the `depth_bin`
+    # value per channel.
+    all_chan_num_range_sample_indices = np.ceil(
+        depth_bin / np.nanmean(np.diff(ds_Sv_copy["depth"], axis=2), axis=(1, 2))
+    ).astype(int)
+
+    # Create list for pooled Sv DataArrays
+    pooled_Sv_list = []
+
+    # Iterate through channels
+    for channel_index in range(len(ds_Sv_copy["channel"])):
+        # Create calibrated Sv DataArray copies and remove values too close to the surface
+        min_range_sample = (ds_Sv["depth"] <= exclude_above).argmin().values
+        chan_Sv = (
+            ds_Sv_copy["Sv"]
+            .copy()
+            .isel(
+                channel=channel_index,
+                range_sample=slice(min_range_sample, len(ds_Sv["range_sample"])),
+            )
+        )
+        chan_pooled_Sv = (
+            ds_Sv_copy["Sv"]
+            .copy()
+            .isel(
+                channel=channel_index,
+                range_sample=slice(min_range_sample, len(ds_Sv["range_sample"])),
+            )
+        )
+
+        # Grab channel-specific number of range sample indices for vertical binning
+        chan_num_range_sample_indices = all_chan_num_range_sample_indices[channel_index]
+
+        # Create pooling size list
+        pooling_size = [(2 * num_side_pings) + 1, (2 * chan_num_range_sample_indices) + 1]
+
+        # Rechunk Sv since `generic_filter` expects a Dask Array
+        chan_Sv = chan_Sv.chunk(chunk_dict)
+
+        # Compute `chan_pooled_Sv` values using dask-image's generic filter
+        chan_pooled_Sv.values = _lin2log(
+            dask_image.ndfilters.generic_filter(
+                chan_Sv.pipe(_log2lin).data,
+                function=np.nanmean if func == "nanmean" else np.nanmedian,
+                size=pooling_size,
+            ).compute()
+        )
+
+        # Remove invalid pooled Sv values that are too close to the edge of the echogram
+        chan_pooled_Sv = chan_pooled_Sv.isel(
+            ping_time=slice(num_side_pings, -num_side_pings),
+            range_sample=slice(chan_num_range_sample_indices, -chan_num_range_sample_indices),
+        )
```

Comment:
i think you mentioned that this is because the wrap-around operation the dask filtering operation does. Is that documented somewhere? Cutting these out would result into some samples not being run through I think.

Seems like an alternative is to pad (using nearest neighbor) the outer rim of the image based on the filter size, and then trim those back to the original size?

Reply:
Yeah, it would result in not filtering out the edge values. 

Under `generic_filter`, there's a parameter called `mode` that does this exact operation: https://image.dask.org/en/latest/dask_image.ndfilters.html.
I think `mode=nearest` or `mode=reflect` may be what we want here. My issue with `nearest` is that if the value at the end is something that is an obvious outlier, then it could really mess up the mean/median aggregation calculation and wouldn't be representative of the region. With `reflect`, we use many more surrounding values, and even if it is repeating some of the values that are already within the aggregation, it will still be more representative of the region as a whole (than that of `nearest`).

Reply:
Using `mode=reflect` after discussion today